### PR TITLE
Git - implement cancellation for updating the model

### PR DESCRIPTION
**Pull request by lszomoru**

### CODE REVIEW

File: extensions/git/src/git.ts [hunk 1]
```diff
@@ -1993,6 +1998,24 @@ export class Repository {
 			child.on('error', cpErrorHandler(e));
 			child.on('exit', onExit);
 		});
+
+		if (opts?.cancellationToken) {
+			const cancellationPromise = new Promise<{ status: IFileStatus[]; statusLength: number; didHitLimit: boolean }>((_, e) => {
+				onceEvent(opts.cancellationToken!.onCancellationRequested)(() => {
```
Comment (joaomoreno) [inline extensions/git/src/git.ts:2004]: We should dispose this listener if child ends up dying before the event gets fired.

File: extensions/git/src/git.ts [hunk 2]
```diff
@@ -1993,6 +1998,24 @@ export class Repository {
...
+					try {
+						child.kill();
+					} catch (err) {
+						// noop
+					}
+
+					e(new GitError({ message: 'Cancelled' }));
```
Comment (joaomoreno) [inline extensions/git/src/git.ts:2011]: Should this rather be a `new CancellationError` instead? We might have to create such a concept.

File: extensions/git/src/git.ts [hunk 3]
```diff
@@ -550,8 +550,8 @@ export class Git {
 
 		if (options.log !== false) {
 			const startTime = Date.now();
-			child.on('exit', (_) => {
-				this.log(`> git ${args.join(' ')} [${Date.now() - startTime}ms]\n`);
+			child.on('exit', (code) => {
+				this.log(`> git ${args.join(' ')} [${Date.now() - startTime}ms]${code === null ? ' (cancelled)' : ''}\n`);
```
Comment (joaomoreno) [inline extensions/git/src/git.ts:554]: Not sure the following assertion is true: if the code is `null`, the is means that the child was killed by us. There could be other reasons for the code to be `null`. There could also be other reasons for the child to be killed.

**Review by joaomoreno — CHANGES_REQUESTED** [review]

File: extensions/git/src/git.ts (hunk 1 above)
Comment (lszomoru) [reply extensions/git/src/git.ts]: Done.

File: extensions/git/src/git.ts (hunk 3 above)
Comment (lszomoru) [reply extensions/git/src/git.ts]: I have switched to using the `killed` property. Another idea that I considered was to pass a custom signal when killing the process as part of cancellation but if I read the documentation correctly, on Windows the signal is being ignored so I am not sure that we would be able to access that. Let me know if you have a better suggestion. Thanks!

**Review by joaomoreno — APPROVED** [review]

**Pull request merged**